### PR TITLE
Slack 

### PR DESCRIPTION
**Pull request by Jonathanko52**
Automatically sends zoom link before a scheduled event in the database. Requires someone on the Slack Administration to set up the app in the Slack API and an updated .env file.

### CODE REVIEW

File: backend/routers/slack.router.js
Comment (nickbeaird) [inline backend/routers/slack.router.js]: I am a little confused by this. Are you aiming to GET or POST on this route?

File: backend/routers/slack.router.js
Comment (nickbeaird) [inline backend/routers/slack.router.js]: I really like that you are creating a separate function for scheduling a message, but do we need to hardwire function to the api res object? I feel like we can either make this function not be Express Middleware and be only a function that we send the Events to, or maybe review if we can turn this into an Express endpoint and not be Middleware.

File: backend/routers/slack.router.js
Comment (nickbeaird) [inline backend/routers/slack.router.js]: I think that this relates to my statement above. I think that we are using a lot of space writing to the response object. We can likely use this logic inside the other call without.

File: backend/routers/slack.router.js
Comment (nickbeaird) [inline backend/routers/slack.router.js]: I am not sure on what the performance of this query is, but we can do this in one call versus multiple. We can send a query to the DB to only return values where the date is greater than the date you want. The DB will then return a smaller query. We would not need to iterate over the longer array.

File: backend/workers/slackbot.js
Comment (nickbeaird) [inline backend/workers/slackbot.js]: Please move any hard coded values to the .env file.

File: client/package-lock.json
Comment (nickbeaird) [inline client/package-lock.json]: We are using yarn-lock files now rather than package.lock.json. Please fix this. 

1. Delete your node_modules
2. Delete your package.json files
3. Yarn install on all the folders 
4. Commit your yarn-lock file.

File: client/src/pages/Home.js
Comment (nickbeaird) [inline client/src/pages/Home.js]: Do we need to hardcode checkInReady on all API calls?

File: client/src/pages/ProjectLeaderDashboard.js
Comment (nickbeaird) [inline client/src/pages/ProjectLeaderDashboard.js]: I would review the function that contains this change and see if we can not mutate outside data, but rather create copies.

File: routers/events.router.js
Comment (nickbeaird) [inline routers/events.router.js]: I don't feel like this is a protected endpoint. We are allowing the end user to put in any query if they have entered checkInReady === true. I think that we can add some better validation logic on this using middleware, rather than checking on this value here and passing in anything on the DB call if not.

File: routers/events.router.js
Comment (nickbeaird) [inline routers/events.router.js]: Our GET query above has more validation than our POST.

File: backend/app.js
Comment (alex-anakin) [inline backend/app.js]: @Jonathanko52 Keep it on one line, please.

File: backend/routers/slack.router.js
Comment (alex-anakin) [reply backend/routers/slack.router.js]: @Jonathanko52 In general, GET is for requesting any data, POST is for creating something. Please use POST for the endpoint because its purpose looks like "creating a new schedule for the message".
@nickbeaird The code review best practices recommend providing reason what caused the comment in case of opening a discussion. Or strict recommendation in case if you understand what is wrong.

File: backend/routers/slack.router.js
Comment (alex-anakin) [reply backend/routers/slack.router.js]: @nickbeaird Looks like the PR author follows the current development style of the application. Don't forget about the existing legacy code. If you'd like to refactor this please create a followup task that can be done as part of resolving the application technical debt.
@Jonathanko52 skip this comment.

File: backend/routers/slack.router.js
Comment (alex-anakin) [reply backend/routers/slack.router.js]: @nickbeaird Please take a look at my response to your previous comment.
@Jonathanko52 skip this comment.

File: backend/routers/slack.router.js
Comment (alex-anakin) [reply backend/routers/slack.router.js]: @nickbeaird There are no multiple calls here. There is still the one call but it queries all events from the DB.
@Jonathanko52 In other words please add a condition for a date into the `find` method.

File: backend/workers/slackbot.js
Comment (alex-anakin) [reply backend/workers/slackbot.js]: @Jonathanko52 I agree with Nick here.

File: client/src/pages/Home.js
Comment (alex-anakin) [reply client/src/pages/Home.js]: @nickbeaird This is a good question but it's better to address it to the legacy codebase. Please create a task for fixing this API endpoint and further client cleanup with a clear description of what should be done.
@Jonathanko52 skip this comment.

File: client/src/pages/AdminDashboard.js
Comment (alex-anakin) [inline client/src/pages/AdminDashboard.js]: @Jonathanko52 This file should not be changed in the scope of the task. Please revert all changes for `client/src/pages/AdminDashboard.js` file.

File: client/src/pages/ProjectLeaderDashboard.js
Comment (alex-anakin) [reply client/src/pages/ProjectLeaderDashboard.js]: @nickbeaird The data from DB contains some legacy records that should be processed before further use. Original outside data is useless because it breaks rendering.
@Jonathanko52 skip this comment.

File: routers/events.router.js
Comment (alex-anakin) [reply routers/events.router.js]: @nickbeaird Protecting endpoints is a good question but it's better to address it to the legacy codebase. Please create a task for fixing this API issue and further client cleanup with a clear description of what should be done.
@Jonathanko52 skip this comment.

File: routers/events.router.js
Comment (alex-anakin) [reply routers/events.router.js]: @nickbeaird Of course. See my previous comment.
@Jonathanko52 skip this comment.

**Review by alex-anakin — CHANGES_REQUESTED** [review]
@Jonathanko52 Please resolve my comments.

File: backend/routers/slack.router.js
Comment (nickbeaird) [reply backend/routers/slack.router.js]: @alex-anakin, this function has been rewritten from an enpoint (req, res) and made into middleware (req, res, next). That makes this call now multiple calls, as it will be passed onto the next call. https://expressjs.com/en/guide/using-middleware.html.

File: backend/routers/slack.router.js
Comment (nickbeaird) [reply backend/routers/slack.router.js]: For both @alex-anakin  and @Jonathanko52 , I asked the question, so that I could ask some follow up questions. I feel like you have both a goal of having a GET route with the findCurrentEvents method, as well as a post with scheduleMessage. I think that both can be true.

File: backend/routers/slack.router.js
Comment (nickbeaird) [reply backend/routers/slack.router.js]: @alex-anakin, I respectfully must disagree. The use of Express middleware was only recently added by me, so there has not been any prior use of middleware to stipulate this as a current development style

From the Express docs, middleware is often used for validation of endpoints (https://expressjs.com/en/guide/using-middleware.html). I will restate that this method seems to be a really great method that could be used by a variety of endpoints. I don't think that using middleware is the most sustainable coding practice moving forward, and that pattern of mutating the response object in middleware endpoints can really make for challenging debugging later. I am suggesting that we use this method as either an endpoint (i.e (req, res)), or as a helper method.

File: backend/routers/slack.router.js
Comment (nickbeaird) [reply backend/routers/slack.router.js]: @alex-anakin, I must again respectfully disagree. 

This comment that I made relates to the above statement in regards to the POST middleware, but is slightly different. The method here is findCurrentEvents, and stands out to me as a GET endpoint. I would assume that we are asking to return all events that are current and that could be an updated GET. This method however is using Express middleware with the use of the next variable in the method (see link below). The docs for Express Middleware suggest using Middleware for when you need to chain methods together across API endpoints, such as setting headers across a wide array of endpoints, or adding in verification for routes. This allows you to isolate your verification of endpoints away from the actual logic. I would again state that this method and the logic in it are best suited to be pulled together in the endpoint, and not used as Express middleware across files. 

https://expressjs.com/en/guide/using-middleware.html

File: client/src/pages/Home.js
Comment (nickbeaird) [reply client/src/pages/Home.js]: @alex-anakin, this is a new method, so it is not legacy code. I defer to you on this one though. I can only again suggest that we try and improve the code that we are adding, rather than adding it to future work.

**Comment by Jonathanko52** [comment]
Changes have been addressed. Waiting for review.

**Pull request closed without merging**
